File: algorithm/study/Core-0B-Graph-Tree/1167/python/src/solution.py
import sys
from collections import deque
input = sys.stdin.readline

def main():
    sys.setrecursionlimit(200000)
    v = int(input())
    adj = [[] for _ in range(v + 1)]
    for _ in range(v):
        data = list(map(int, input().split()))
        node = data[0]
        i = 1
        while data[i] != -1:
            neighbor, weight = data[i], data[i + 1]
            adj[node].append((neighbor, weight))
            i += 2

    def bfs(start):
        dist = [-1] * (v + 1)
        dist[start] = 0
        q = deque([start])
        far_node, far_dist = start, 0
        while q:
            u = q.popleft()
            for nv, w in adj[u]:
                if dist[nv] == -1:
                    dist[nv] = dist[u] + w
                    q.append(nv)
                    if dist[nv] > far_dist:
                        far_dist = dist[nv]
                        far_node = nv
        return far_node, far_dist

    u, _ = bfs(1)
    _, diameter = bfs(u)
    print(diameter)

if __name__ == "__main__":
    main()

File: algorithm/study/Core-0B-Graph-Tree/1167/cpp/src/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int V;
vector<pair<int,int>> adj[100001];
pair<int,long long> bfs(int s){
    vector<long long> dist(V+1, -1);
    queue<int> q;
    dist[s]=0; q.push(s);
    int far_node=s; long long far_dist=0;
    while(!q.empty()){
        int u=q.front(); q.pop();
        for(auto [v,w]:adj[u]){
            if(dist[v]==-1){
                dist[v]=dist[u]+w;
                q.push(v);
                if(dist[v]>far_dist){ far_dist=dist[v]; far_node=v; }
            }
        }
    }
    return {far_node, far_dist};
}
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin >> V;
    for(int i=0;i<V;i++){
        int node; cin >> node;
        int x;
        while(cin >> x && x != -1){
            int w; cin >> w;
            adj[node].push_back({x,w});
        }
    }
    auto [u,_] = bfs(1);
    auto [__, diameter] = bfs(u);
    cout << diameter << '\n';
    return 0;
}
